### PR TITLE
*: support reload tls used by mysql protocol in place

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

ref #14666

preliminary support reload tls used by mysql protocol

this PR doesn't try to full support [mysql's dynamic modify "ssl_ca/ssl_key/ssl_cert" value](https://dev.mysql.com/doc/refman/8.0/en/alter-instance.html#alter-instance-reload-tls), but can reload tls used old file path specified by old "ssl_ca/ssl_key/ssl_cert" value(so ssl_cert/ssl_ca/ssl_key  keep read-only after this PR).

so user can:

1. start TiDB with ssl-ca, ssl-key and ssl-cert config like https://pingcap.com/docs/stable/reference/security/cert-based-authentication/#install-openssl
2. replace new file specified in ssl-ca, ssl-key and ssl-cert
3. use super user(here need new priv in following pr) to execute `alter instance reload tls`

then all new db connection will use new cert file, old connection will keep work just like mysql does

### What is changed and how it works?

- extract common method `LoadTLSCertificates`
- make `server.tlsConfig` can be atomic swap
- let alter instance reload tls do reload

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - impl change

Side effects

 - n/a

Related changes

 - 4.0 only

Release note

 - support reload tls used by mysql protocol in place.

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/14749)

### CODE REVIEW

File: executor/simple.go [hunk 1]
```diff
@@ -1098,6 +1101,26 @@ func (e *SimpleExec) executeFlush(s *ast.FlushStmt) error {
...
+			variable.SysVars["ssl_key"].Value,
+			variable.SysVars["ssl_cert"].Value,
+		)
+		if err != nil {
+			if !s.NoRollbackOnError {
+				return err
+			}
+			logutil.BgLogger().Warn("reload tls fail but keep work without TLS due to 'no rollback on error'")
+		}
+		sm.UpdateTLSConfig(tlsCfg)
```
Comment (jackysp) [inline executor/simple.go]: Should we `UpdateTLSConfig` when `NoRollbackOnError`?

File: executor/simple.go [hunk 2]
```diff
@@ -1098,6 +1101,26 @@ func (e *SimpleExec) executeFlush(s *ast.FlushStmt) error {
...
+		tlsCfg, err := util.LoadTLSCertificates(
+			variable.SysVars["ssl_ca"].Value,
+			variable.SysVars["ssl_key"].Value,
+			variable.SysVars["ssl_cert"].Value,
+		)
+		if err != nil {
+			if !s.NoRollbackOnError {
+				return err
+			}
+			logutil.BgLogger().Warn("reload tls fail but keep work without TLS due to 'no rollback on error'")
```
Comment (jackysp) [inline executor/simple.go]:
```suggestion
			logutil.BgLogger().Warn("reload TLS failed, but keep working without TLS due to 'no rollback on error'")
```

File: executor/simple.go [hunk 3]
```diff
@@ -1098,6 +1101,26 @@ func (e *SimpleExec) executeFlush(s *ast.FlushStmt) error {
 	return nil
 }
 
+func (e *SimpleExec) executeAlterInstance(s *ast.AlterInstanceStmt) error {
+	if s.ReloadTLS {
+		logutil.BgLogger().Info("execute reload tls")
```
Comment (jackysp) [inline executor/simple.go]: Not sure if it is necessary or not.

File: server/server.go [hunk 4]
```diff
@@ -209,16 +208,23 @@ func NewServer(cfg *config.Config, driver IDriver) (*Server, error) {
 		clients:           make(map[uint32]*clientConn),
 		stopListenerCh:    make(chan struct{}, 1),
 	}
-	s.loadTLSCertificates()
+
+	tlsConfig, err := util.LoadTLSCertificates(s.cfg.Security.SSLCA, s.cfg.Security.SSLKey, s.cfg.Security.SSLCert)
+	if err != nil {
+		logutil.BgLogger().Warn("secure connection is not enabled")
```
Comment (jackysp) [inline server/server.go]: I think we should return the error.

Comment (jackysp) [inline server/server.go]: I think we should return the error here, but the old implement doesn't return an error either. It is weird.

File: executor/simple.go (hunk 1 above)
Comment (lysu) [reply executor/simple.go]: yes 

> If the optional NO ROLLBACK ON ERROR clause is given and a new context cannot be created, rollback does not occur. Instead, a warning is generated and SSL is disabled for new connections.

MySQL will keep work with nil ssl config

File: server/server.go (hunk 4 above)
Comment (lysu) [reply server/server.go]: +1 we should return error here

File: executor/simple.go (hunk 3 above)
Comment (lysu) [reply executor/simple.go]: maybe better keep this log, reload tls is critial operation, we should keep this log to help diagnose, we also add s.NoRollbackOnError value to this log in new commit

**Comment by tiancaiamao** [comment]
```
[2020-02-19T06:31:54.630Z] ----------------------------------------------------------------------
[2020-02-19T06:31:54.630Z] FAIL: tidb_test.go:468: tidbTestSuite.TestReloadTLS
[2020-02-19T06:31:54.630Z] 
[2020-02-19T06:31:54.630Z] tidb_test.go:524:
[2020-02-19T06:31:54.630Z]     c.Assert(err, IsNil)
[2020-02-19T06:31:54.631Z] ... value tls.RecordHeaderError = tls.RecordHeaderError{Msg:"first record does not look like a TLS handshake", RecordHeader:[5]uint8{0x17, 0x0, 0x0, 0x2, 0xff}, Conn:(*net.TCPConn)(0xc003beb3b0)} ("tls: first record does not look like a TLS handshake")
[2020-02-19T06:31:54.631Z] 
[2020-02-19T06:31:54.631Z] 
[2020-02-19T06:31:54.631Z] ----------------------------------------------------------------------
```

**Comment by kolbe** [comment]
I'd like to see this backported to 3.1 and maybe 3.0.

File: server/server_test.go [hunk 5]
```diff
@@ -17,6 +17,7 @@ import (
 	"database/sql"
 	"encoding/json"
 	"fmt"
+	"github.com/pingcap/errors"
```
Comment (jackysp) [inline server/server_test.go]: fmt

**Comment by lysu** [comment]
@jackysp @tiancaiamao  finally it passes all test,  PTAL tks~

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**